File: booksy-admin/src/utils/person-name.ts
/**
 * What counts as a person's real name (the same rule as booksy-frontend's core/utils/person-name).
 *
 * A person signed up by phone is stored as «مشتری 9384444636» / «ارائه‌دهنده 9123135143» — the word for their side
 * of the marketplace, then their phone's national number — until they give a name. Production QA 2026-09-23 found
 * the apps printing that as a name: "the number must never be written anywhere". Mirrors the API's `PersonName`:
 * a run of 7+ digits (any spacing, Latin or Persian digits) is a phone and never part of a name; a digits-only
 * surname is not a surname; the bare placeholder word is not a name.
 */

const PLACEHOLDER_WORDS = new Set(['مشتری', 'ارائه‌دهنده', 'ارائه دهنده'])

/** A run of digits this long is a phone number, however it is spaced; shorter ones can be part of a name. */
const PHONE_DIGITS = 7

const DIGIT = /[0-9۰-۹٠-٩]/g
const DIGIT_GROUP = /^[0-9۰-۹٠-٩+\-().\/]+$/

const digitCount = (token: string): number => token.match(DIGIT)?.length ?? 0
const isDigitGroup = (token: string): boolean => DIGIT_GROUP.test(token) && digitCount(token) > 0

/** A name held as one string with any phone number taken out; null when nothing real is left. */
export function personNameOrNull(name?: string | null): string | null {
  if (!name || !name.trim()) return null

  // Split on real whitespace only: the ZWNJ inside «ارائه‌دهنده» is part of the word.
  const kept: string[] = []
  let run: string[] = []
  const endRun = () => {
    // Consecutive digit groups ("0912 313 5143") are one number; drop it when it is a phone.
    if (run.reduce((n, t) => n + digitCount(t), 0) < PHONE_DIGITS) kept.push(...run)
    run = []
  }

  for (const token of name.trim().split(/\s+/)) {
    if (isDigitGroup(token)) {
      run.push(token)
      continue
    }
    endRun()
    kept.push(token)
  }
  endRun()

  const result = kept.join(' ')
  return result === '' || PLACEHOLDER_WORDS.has(result) ? null : result
}

/** The person's real name from their two parts, or null when they have none. */
export function realNameOrNull(firstName?: string | null, lastName?: string | null): string | null {
  const first = personNameOrNull(firstName) ?? ''
  const rawLast = lastName?.trim() ?? ''
  // The placeholder's surname is the phone's national number: a surname made of digits is not one.
  const last = isDigitGroup(rawLast) ? '' : (personNameOrNull(rawLast) ?? '')
  const full = `${first} ${last}`.trim()
  return full === '' ? null : full
}

File: asan-rezerve-provider-app/lib/core/utils/person_name.dart
/// Splits and joins a person's name for the single "full name" field.
///
/// The UI asks for one full name, but the backend stores — and requires — first and last
/// name separately (SaveStep3LocationCommandHandler rejects either being blank). So the
/// full name is split before it is sent.
///
/// **Rule: split on the LAST run of whitespace.** Everything before it is the first name,
/// the final token is the last name. Chosen for Persian:
/// - compound first names are commonly written with a real space — «محمد علی رضایی» →
///   first «محمد علی», last «رضایی» (splitting on the FIRST space would get this wrong);
/// - compound surnames are conventionally joined with a ZERO WIDTH NON-JOINER (نیم‌فاصله,
///   U+200C), which is not whitespace, so «علی حسینی‌نژاد» keeps «حسینی‌نژاد» whole.
///
/// The app always displays `first + " " + last`, so even a surname written with a real
/// space (a mis-split) still reconstructs exactly what the user typed.
class PersonName {
  PersonName._();

  // Regular whitespace only. U+200C (ZWNJ) is a format character, not whitespace, and must
  // never be treated as a separator.
  static final RegExp _whitespace = RegExp(r'[\s ]+');

  /// Returns `(first, last)`, or `null` when the input has fewer than two words (there is
  /// no last name to send, and the backend would reject the request).
  static ({String first, String last})? split(String fullName) {
    final tokens = fullName
        .trim()
        .split(_whitespace)
        .where((t) => t.isNotEmpty)
        .toList();
    if (tokens.length < 2) return null;
    return (
      first: tokens.sublist(0, tokens.length - 1).join(' '),
      last: tokens.last,
    );
  }

  /// Rebuilds the single field from stored parts (e.g. when a saved draft is restored).
  static String join(String first, String last) =>
      [first.trim(), last.trim()].where((p) => p.isNotEmpty).join(' ');

  // ---- What counts as a real name (mirrors PersonName on the server) ----
  //
  // Phone sign-in names a person «ارائه‌دهنده 9123135143» — the word for their side of the marketplace, then their
  // phone's national number — until they give a name. Production QA 2026-09-23 found the salon app showing that,
  // and the bare phone, as the owner's name: "the number must never be written anywhere".

  static const _placeholderWords = {'مشتری', 'ارائه‌دهنده', 'ارائه دهنده'};

  /// A run of digits this long is a phone number, however it is spaced; shorter ones can be part of a name.
  static const _phoneDigits = 7;

  static final RegExp _digit = RegExp(r'[0-9\u06F0-\u06F9\u0660-\u0669]');
  static final RegExp _digitGroup = RegExp(r'^[0-9\u06F0-\u06F9\u0660-\u0669+\-().\/]+$');

  static bool _isDigitGroup(String token) =>
      _digitGroup.hasMatch(token) && _digit.hasMatch(token);

  /// The person's real name from their two parts, or null when they have none.
  static String? realOrNull(String? firstName, String? lastName) {
    final first = sanitize(firstName) ?? '';
    // The placeholder's surname is the phone's national number: a surname made of digits is not one.
    final rawLast = lastName?.trim() ?? '';
    final last = _isDigitGroup(rawLast) ? '' : sanitize(rawLast) ?? '';
    final full = join(first, last);
    return full.isEmpty ? null : full;
  }

  /// A name held as one string with any phone number taken out; null when what is left is nothing or the bare
  /// placeholder word.
  static String? sanitize(String? name) {
    if (name == null || name.trim().isEmpty) return null;

    final kept = <String>[];
    final run = <String>[];
    void endRun() {
      // Consecutive digit groups ("0912 313 5143") are one number; drop it when it is a phone.
      final digits = run.fold<int>(0, (n, t) => n + _digit.allMatches(t).length);
      if (digits < _phoneDigits) kept.addAll(run);
      run.clear();
    }

    for (final token in name.trim().split(_whitespace)) {
      if (token.isEmpty) continue;
      if (_isDigitGroup(token)) {
        run.add(token);
        continue;
      }
      endRun();
      kept.add(token);
    }
    endRun();

    final result = kept.join(' ');
    return result.isEmpty || _placeholderWords.contains(result) ? null : result;
  }
}
